### ISSUE TITLE
[exporter/loki] Atributes targeted in hint not converted to labels

### ISSUE THREAD

**Issue by gillg** [open-telemetry/opentelemetry-collector-contrib#22038]
### Component(s)

exporter/loki, pkg/translator/loki

## Description
For a completely unknonw reason some attributes pushed via OTLP and targeted in `loki.attribute.labels` are not converted to labels.
As a converted label is removed from the attributes map we can also see the problem because at the end the attributes are still present as attributes in loki.

I tried to add a static value to all the records via an attribute processor and in that case the records which don't have the attribute initaly are finaly converted correctly to labels.

## Steps to Reproduce
```
receivers:
 otlp:
  protocols:
    grpc:
      endpoint: '[::]:4317' # logs comming from a custom applications

 filelog: [...] # parse docker logs

processors:
  resource/loki:
     attributes:
      # Define Loki format
      - action: insert
        key: loki.format
        value: json

      # Define Loki resource
      - action: insert 
        key: loki.resource.labels
        value: >-
          container.name, net.host.name, service.name, service.version,
          instance, k8s.container.name, k8s.namespace.name, exporter

      - action: insert
        key: instance
        from_attribute: net.host.name

 attributes/loki:
  actions:
    # Define Loki attributes
    - action: insert
      key: loki.format
      value: json

    - action: insert
      key: loki.attribute.labels
      value:
       # Type array used for lisibility
        - fluent.tag
        - fluent.filename
        - fluentd_container_name
        - container_name
        - elastic.cluster.name
        - elastic.log.severity
        - elastic.log.type
        - elastic.node.name
        - event.domain
        - k8s.resource.name
        - winevent.computername
        - winevent.severity
        - winevent.source
        - instance
        - cortex_env
        - webserver.host
        - webserver.computername
        - webserver.sitename
        - cortex_location
        - cortex_service
        - environment
        - ab_customer
        - severity
        - app
        - ab.global.environment
        - ab.cortex.log.level
        - ab.global.application.name
        - ab.cortex.tenant.location
        - ab.cortex.tenant.state
        - log.iostream

    # Test static value
    - action: insert
      key: ab.cortex.tenant.state
      value: fake # this value is added to all the logs if they don't already have an attribute "ab.cortex.tenant.state"

    # Attribute remapping for backward compatibility
    - action: insert
      key: instance
      from_attribute: elastic.node.name

    - action: insert
      key: instance
      from_attribute: winevent.computername

    - action: insert
      key: cortex_env
      from_attribute: ol.cortex.tenant.state

    - action: insert
      key: cortex_location
      from_attribute: ol.cortex.tenant.location

    - action: insert
      key: cortex_service
      from_attribute: ol.cortex.service

    - action: insert
      key: environment
      from_attribute: ol.global.environment

    - action: insert
      key: ab_customer
      from_attribute: ol.cortex.tenant.id

    - action: insert
      key: severity
      from_attribute: ol.cortex.log.level

    - action: insert
      key: app
      from_attribute: ol.global.application.name

    - action: insert
      key: cortex_user_source
      from_attribute: ol.cortex.user.source

```

## Expected Result

Having all the hint attributes removed from application logs in Loki (by the way that could be an option to keep them in the log entry), and have a loki label with their value.

## Actual Result

Log working:
```
{
  "body": "level=info ts=2023-05-17T12:12:01.529656354Z caller=metrics.go:149 component=querier org_id=fake latency=fast query=\"{exporter=\\\"OTLP\\\"}\" query_hash=470100140 query_type=limited range_type=range length=12m0.645s start_delta=12m1.529648236s end_delta=884.648427ms step=1s duration=358.174844ms status=200 limit=1000 returned_lines=1000 throughput=41MB total_bytes=15MB total_entries=1000 store_chunks_download_time=333.467692ms queue_time=0s subqueries=1 cache_chunk_req=4 cache_chunk_hit=0 cache_chunk_bytes_stored=145 cache_chunk_bytes_fetched=0 cache_chunk_download_time=9.9µs cache_index_req=0 cache_index_hit=0 cache_index_download_time=0s cache_result_req=0 cache_result_hit=0 cache_result_download_time=0s",
  "attributes": {
    "log.file.path": "/var/log/pods/loki_loki-loki-distributed-querier-2_ad7d1674-ea78-4db9-a7d0-38894f1c824a/querier/0.log",
    "logtag": "F",
    "time": "2023-05-17T12:12:01.623349807Z"
  },
  "resources": {
    "k8s.container.restart_count": "0",
    "k8s.pod.name": "loki-loki-distributed-querier-2",
    "k8s.pod.uid": "ad7d1674-ea78-4db9-a7d0-38894f1c824a"
  }
}
Labels:
cortex_env | fake
exporter | OTLP
k8s_container_name | querier
k8s_namespace_name | loki
log_iostream | stderr
ab_cortex_tenant_state | fake
```

Log not working, containing attributes natively coming from OTLP receiver
```
{
  "body": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  "severity": "Debug",
  "attributes": {
    "app": "Cortex",
    "cortex_env": "xxxxx",
    "cortex_location": "xxx",
    "cortex_service": "xxxxx",
    "environment": "Test",
    "net.host.name": "xxxxxxxx",
    "net.peer.ip": "xxxxxxxx",
    "ab.cortex.app_name": "xxxxxxxx",
    "ab.cortex.base_url": "xxxxxxxxx",
    "ab.cortex.log.function_name": "xxxxx",
    "ab.cortex.log.job_number": "xxxxxx",
    "ab.cortex.log.level": "Verbose",
    "ab.cortex.log.module_name": "xxxxxx",
    "ab.cortex.log.process_id": "xxxxx",
    "ab.cortex.log.raw_datetime": "5/17/2023 12:12:01 PM",
    "ab.cortex.log.thread_id": "120",
    "ab.cortex.log.timestamp": "1684325521628214900",
    "ab.cortex.parameters.profile": "xxxxxxxxx",
    "ab.cortex.parameters.session": "xxxxxx",
    "ab.cortex.service": "xxxx",
    "ab.cortex.tenant.id": "xxxxx",
    "ab.cortex.tenant.location": "xxxx",
    "ab.cortex.tenant.name": "xxxxx",
    "ab.cortex.tenant.state": "Standard",
    "ab.cortex.user.id": "xxxxxxx",
    "ab.global.application.name": "Cortex",
    "ab.global.application.version.name": "xxxxxxxxx",
    "ab.global.client.name": "xxxxx",
    "ab.global.environment": "xxxx",
    "ab_customer": "xxxxx",
    "severity": "Verbose"
  }
}
Labels:
exporter | OTLP
level | DEBUG
```

### Collector version

0.77.0

## Environment
Linux / Kubernetes / Containerd

**Comment by mar4uk** [comment]
Could you please add 
```
logging: 
   verbosity: detailed
``` 
exporter and send the LogRecord you see in the logging exporter output? It will help me to understand the structure of the LogRecord you try to send to Loki.

I didn't understand the actual result section. Why there are 2 cases? Is the first one correct and the second one is not?

**Comment by gillg** [comment]
Sorry for the lack of clarity, yes there is two cases the first is working and the second is not. One is coming from OTLP and the other is coming from filelog receiver.

In addition other same examples but formatted from the logging exporter.

This is a "failing" record (the attributes are not converted to labels):
```
LogRecord #50
ObservedTimestamp: 1970-01-01 00:00:00 +0000 UTC
Timestamp: 2023-05-22 18:17:10.1740092 +0000 UTC
SeverityText: Information
SeverityNumber: Info(9)
Body: Str(Apply Long Running Background Status Updates is returning 'Job executed correctly' (executed in 487ms))
Attributes:
     -> ab.global.application.name: Str(Cortex)
     -> ab.global.application.version.name: Str(xxxxxxx)
     -> ab.global.client.name: Str(xxxxx)
     -> ab.global.environment: Str(Test)
     -> ab.cortex.app_name: Str(xxxxxxxxx)
     -> ab.cortex.base_url: Str(xxxxxxxxxx)
     -> ab.cortex.service: Str(Agent)
     -> ab.cortex.tenant.name: Str(xxxxxxxx)
     -> ab.cortex.tenant.id: Str(xxxxxx)
     -> ab.cortex.tenant.location: Str(xxxxxx)
     -> net.host.name: Str(xxxxxx)
     -> net.peer.ip: Str(127.0.0.1)
     -> net.host.port: Str(443)
     -> http.url: Str(xxxxxxx)
     -> http.method: Str(GET)
     -> http.status_code: Str(200)
     -> http.scheme: Str(https)
     -> http.client_ip: Str(xxxxxxx)
     -> ab.cortex.user.id: Str(xxxxxx)
     -> ab.cortex.parameters.profile: Str(xxxxxxxx)
     -> ab.cortex.parameters.session: Str(xxxxxxxx
)
     -> ab.cortex.tenant.state: Str(Standard)
     -> ab.cortex.log.level: Str(Information)
     -> ab.cortex.log.function_name: Str(Run)
     -> ab.cortex.log.module_name: Str(xxxxxxxxxx)
     -> ab.cortex.log.thread_id: Str(52)
     -> ab.cortex.log.process_id: Str(57552)
     -> ab.cortex.log.job_number: Str(xxxxxx)
     -> ab.cortex.log.job_name: Str(Apply Long Running Background Status Updates)
     -> ab.cortex.log.job_function_name: Str(xxxxxxxxxx)
     -> ab.cortex.log.raw_datetime: Str(5/22/2023 6:17:10 PM)
     -> loki.attribute.labels: Slice(["fluent.tag","fluent.filename","fluentd_container_name","container_name","elastic.cluster.name","elastic.log.severity","elastic.log.
type","elastic.node.name","event.domain","k8s.resource.name","winevent.computername","winevent.severity","winevent.source","instance","cortex_env","webserver.host","webse
rver.computername","webserver.sitename","cortex_location","cortex_service","environment","ab_customer","severity","app","ab.global.environment","ab.cortex.log.level","ab.
global.application.name","ab.cortex.tenant.location","ab.cortex.tenant.state","cortex_user_source","cortex_user_login","log.iostream"])
     -> cortex_env: Str(Standard)
     -> cortex_location: Str(xxxxxx)
     -> cortex_service: Str(Agent)
     -> environment: Str(Test)
     -> ab_customer: Str(x)xxxx
     -> severity: Str(Information)
     -> app: Str(Cortex)
Trace ID:
Span ID:
Flags: 0
```
=> Resulting labels:
| exporter | OTLP |
| level | DEBUG |


This is a "working" record (the attributes are converted to labels):
```
LogRecord #1
ObservedTimestamp: 1970-01-01 00:00:00 +0000 UTC
Timestamp: 2023-05-22 18:17:08 +0000 UTC
SeverityText:
)ody: Str(      "ProcdumpArgs": "-accepteula -ma -r -m 15000 -s 90"
Attributes:
     -> fluent.tag: Str(docker-xxxxxxxxx)
     -> container_id: Str(e4fdc26f68efc03d1d04359eca2ef42e15bde73e7d8a2cc973460eed3c320541)
     -> container_name: Str(xxxxxxxx)
     -> source: Str(stdout)
     -> loki.attribute.labels: Slice(["fluent.tag","fluent.filename","fluentd_container_name","container_name","elastic.cluster.name","elastic.log.severity","elastic.log.
type","elastic.node.name","event.domain","k8s.resource.name","winevent.computername","winevent.severity","winevent.source","instance","cortex_env","webserver.host","webse
rver.computername","webserver.sitename","cortex_location","cortex_service","environment","ab_customer","severity","app","ab.global.environment","ab.cortex.log.level","ab.
global.application.name","ab.cortex.tenant.location","ab.cortex.tenant.state","cortex_user_source","cortex_user_login","log.iostream"])
     -> ab.cortex.tenant.state: Str(fake)
     -> cortex_env: Str(fake)
Trace ID:
Span ID:
Flags: 0
```
=> Resulting labels
|container_name | xxxxx |
| cortex_env | fake |
| exporter | OTLP |
| fluent_tag | docker-xxxxxx |
| ol_cortex_tenant_state | fake |

**Comment by gillg** [comment]
Root cause found !

I don't understand yet why but I got a doubt on this code :
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/loki/logs_to_loki.go#LL288C1-L305C2

So To avoid the "merge" of hints based on `log.Attributes().PutStr(hintAttributes, fmt.Sprintf("%s,%s", value.AsString(), levelAttributeName))`

If added this processor in my config:
```yaml
      transform:
        log_statements:
          - context: log
            statements:
              - set(severity_number, SEVERITY_NUMBER_UNSPECIFIED)
```

And now everything works as expected, all the targeted attributes in hints are converted to labels (except I lost the original severity).

I supect a lot the Sprintf() function not able to merge the severity string with a `Slice`. But in my memories I also tested with hints as a big string coma separated and it was not working either. Maybe I'm wrong.

**Comment by gillg** [comment]
I think I have a better workaround without dataloss but it doesn't work in v < 0.78.0 because the concat function doesn't exist.

```yaml
      transform:
        log_statements:
          - context: log
            statements:
              #- set(severity_number, SEVERITY_NUMBER_UNSPECIFIED)
              - concat(attributes["loki.attribute.labels"], ",")
```

**Comment by gillg** [comment]
By the way, `strings.Contains(value.AsString(), levelAttributeName)` doesn't work either with a slice...

I would open a PR but I'm not confortable with types comparaison in golang and the best way to merge a string or a slice.
But we should definitely add some missing tests with slice hints and defined severity number and pre-existing `level` attribute.

**Comment by jpkrohling** [comment]
This sounds definitely like a bug on our side. @mar4uk, perhaps we should treat string slices differently?

**Comment by mar4uk** [comment]
The issue can be closed. The fix is available in v0.82.0
